[DTensorTestbase] Fix `@with_comms` inactive problem

tsunghsienlee opened the pull request:
Summary:
`with_comms()` is mostly used as a decorator with an optional input argument `eager_init`. The problem of a decorator with input argument is that it has to be used with invocation always, i.e., you have to use as `with_comms()` rather than `with_comms` which majority of the existing usages.

This diff tries to provide a solution such that we could use `with_comms`, `with_comms()`, `with_comms(eager_init=False)`, and `with_comms(eager_init=True)`.

Test Plan: Contbuild & OSS CI

Differential Revision: D65385700




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wz337 reviewed (approved):
LGTM! Thanks for reporting and helping to fix this issue!

tsunghsienlee committed:
[DTensorTestbase] Fix `@with_comms` inactive problem (#139637)

Summary:

`with_comms()` is mostly used as a decorator with an optional input argument `eager_init`. The problem of a decorator with input argument is that it has to be used with invocation always, i.e., you have to use as `with_comms()` rather than `with_comms` which majority of the existing usages.

This diff tries to provide a solution such that we could use `with_comms`, `with_comms()`, `with_comms(eager_init=False)`, and `with_comms(eager_init=True)`.

bypass-github-pytorch-ci-checks

Test Plan: Contbuild & OSS CI

Reviewed By: wz337

Differential Revision: D65385700
